files/Aboot/boot0.j2

```diff
@@ -657,22 +658,53 @@ write_default_cmdline() {
     cmdline_clear
 
     if $in_aboot; then
-        # generate the default kernel parameters for the platform
+        # Generate the default kernel parameters for the platform
         cat /etc/cmdline | sed "/^\(${bootconfigvars// /\|}\|crashkernel\|loglevel\|ignore_loglevel\)\(\$\|=\)/d;/^\$/d" | cmdline_append
-    elif grep -q "$delimiter" /proc/cmdline; then
-        # we are on a recent sonic image using delimiter. extracting the part of the
+    elif grep -q "$delimiter" /proc/cmdline && ! grep -Eq "log_buf_len=.* $delimiter" /proc/cmdline; then
+        # We are on a recent sonic image using delimiter. extracting the part of the
         # cmdline coming from aboot is trivial.
+        # The 2nd part of the condition ensures that the append bug is not present.
+        # When it is it should go through the last case of this if/else block to
+        # regenerate the aboot cmdline
         cat /proc/cmdline | sed -E "s/^(.*) $delimiter .*$/\1/" | tr ' ' '\n' | cmdline_append
-    else
-        # we are either on SONiC or EOS and the commandline doesn't have a delimiter
-        # for the Aboot part. Take an educated guess at a right delimiter.
-        # Subject to breakage if EOS or SONiC cmdline change.
+    elif grep -q "SWI=" /proc/cmdline; then
+        # We are in EOS and the cmdline doesn't have a delimiter for the aboot part
         cat /proc/cmdline | sed -E 's/^(.*) rw .*$/\1/' | tr ' ' '\n' | cmdline_append
+    else
+        # We are in SONiC and the cmdline doesn't have a delimiter for the aboot part.
+        # sid= should most of the time be the last parameter provided by Aboot
+        cat /proc/cmdline | sed -E 's/^(.* sid=[^ ]+).*$/\1/' | tr ' ' '\n' | cmdline_append
```

Comment:
This would remove some unique params from the cmdline too. I am assuming we only want to remove duplicates. Is this intentional?

```
$ cat temp_cmdline
reboot=p console=ttyS0 acpi=on Aboot=Aboot-norcal7-7.2.0-pcie2x4-6128821 block_flash=pci0000:00/0000:00:1f.2/.*host./target0:0:0/.*$ block_drive=pci0000:00/0000:00:1f.2/.*host./target2:0:0/.*$ net_ma2=pci0000:00/0000:00:01.0/.*$ net_ma1=pci0000:00/0000:00:01.1/.*$ block_usb2=pci0000:00/0000:00:14.0/\(usb3/3-2\|usb4/4-2\)/.*$ block_usb1=pci0000:00/0000:00:14.0/\(usb3/3-3\|usb4/4-5\)/.*$ platform=rook scd.lpc_irq=7 scd.lpc_res_addr=0xb0000000 scd.lpc_res_size=0x10000 sid=Gardena iommu=on intel_iommu=on tsc=reliable pcie_ports=native rhash_entries=1 usb-storage.delay_use=0 reassign_prefmem varlog_size=4096 net.ifnames=0 log_buf_len=1M iommu=on intel_iommu=on tsc=reliable pcie_ports=native rhash_entries=1 usb-storage.delay_use=0 reassign_prefmem varlog_size=4096 net.ifnames=0 log_buf_len=1M cmdline-aboot-end tsc=reliable pcie_ports=native rhash_entries=1 usb-storage.delay_use=0 reassign_prefmem iommu=on intel_iommu=on intel_idle.max_cstate=0 varlog_size=4096 sonic.mode=fixed log_buf_len=1M security=apparmor apparmor=1 rw net.ifnames=0 systemd.unified_cgroup_hierarchy=0 quiet systemd.show_status=auto hwaddr_ma1=98:5d:82:1b:0a:2c root=UUID=2f78dbb3-fbea-40f9-b14d-a47a8ecba076 loop=image-20201231.68/fs.squashfs loopfstype=squashfs
$ 
$ cat temp_cmdline | sed -E 's/^(.* sid=[^ ]+).*$/\1/' | tr ' ' '\n'
reboot=p
console=ttyS0
acpi=on
Aboot=Aboot-norcal7-7.2.0-pcie2x4-6128821
block_flash=pci0000:00/0000:00:1f.2/.*host./target0:0:0/.*$
block_drive=pci0000:00/0000:00:1f.2/.*host./target2:0:0/.*$
net_ma2=pci0000:00/0000:00:01.0/.*$
net_ma1=pci0000:00/0000:00:01.1/.*$
block_usb2=pci0000:00/0000:00:14.0/\(usb3/3-2\|usb4/4-2\)/.*$
block_usb1=pci0000:00/0000:00:14.0/\(usb3/3-3\|usb4/4-5\)/.*$
platform=rook
scd.lpc_irq=7
scd.lpc_res_addr=0xb0000000
scd.lpc_res_size=0x10000
sid=Gardena
$ 
```

Reply:
The whole point of the `write_default_config()` function is to generate `/host/kernel-params-base` and the initial content of `/host/image-XXX/kernel-cmdline`.
It is therefore expected to only output Aboot specific parameter and discard the others.

Reply:
About discarding the params - is there any risk in doing so? What about the params such as  `SONIC_BOOT_TYPE`, which are appended to cmdline by (warm/fast/cold)-boot.

Reply:
Not at all, if you look at the function `write_regular_configs()` which is the one used when booting in non-secureboot mode it does the following:
```
 -> write_common_configs()
     -> write_default_cmdline()
         this one write the Aboot provided parameter only and is the one your comment is on
     -> write_platform_specific_cmdline()
         this write the cmdline parameters required for the current product for the current image
         each image is therefore free to add/remove platform specific parameter (e.g kernel upgrades requiring new params)
     -> write_image_specific_cmdline()
         this write the cmdline parameters for the current image which are product independent (rootfs location, systemd configs, ...)
 -> write_cmdline()
     this functions writes /host/kernel-params-base and /host/image-XXX/kernel-cmdline
```
It is therefore expected that `write_default_cmdline()` only outputs the Aboot specific parameters.
Everything is re-computed at every reboot beside `warm-reboot` and `fast-reboot`
In the `fast-reboot` script with have 2 paths one for secureboot and one for non-secureboot.
In this case it grabs `/host/image-XXX/kernel-cmdline` (generated by sonic-installer install) and append SONIC_BOOT_TYPE
See https://github.com/Azure/sonic-utilities/blob/05c79ef0842deb93e05e3947c4179cb6e0c486e6/scripts/fast-reboot#L334